blog: Picking a database migration tool for Go projects in 2023 (#1321)
ariga/atlas: doc/website/blog/2022-12-01-picking-database-migration-tool.mdx

One of Golang migrate main strengths are the database drivers support and migrations sources.

If your project uses a database driver that is not very popular, chances are that Golang
migrate has a driver for it. For cases where your database is not supported, Golang migrate has
a simple API for defining new database drivers. Databases supported by Golang migrate
include: PostgreSQL, Redshift, Ql, Cassandra, SQLite, MySQL/MariaDB, Neo4j, MongoDB,
Google Cloud Spanner, and more.

Another feature of Golang migrate is the support for different migrations sources, for cases where your
migration scripts resides on custom locations or even remote servers.

### Goose
Goose has a similar approach to Golang migrate, initially was created by [Liam
Staskawicz](https://bitbucket.org/liamstask/) in [2012](https://bitbucket.org/liamstask/goose/src/3e2b32284c219f120a588f1fe4798be671c94dfb/),
[Pressly](https://github.com/pressly) created a fork in 2016 improving the usage, adding support for
migrations in Go, handling cases of migrations out of order and custom schemas for migration
versioning, today Goose have 80 contributors.

> Goose only provides support for 7 database drivers, if your project uses one of the main

Reviewer: ```suggestion
Goose only provides support for 7 database drivers, so if your project uses one of the main
```

[one key issue with Golang migrate](https://github.com/golang-migrate/migrate/issues/510).

### Atlas
Atlas takes a completely different approach to Golang migrate or Goose, while both
tools only focus on proving means of running and maintaining the migration directory, Atlas
takes one step further and actually constructs a graph representing the different database
entities from the migration directory contents, allowing for more complex
scenarios and providing safety for migration operations.

Migrations in Atlas can be defined in two ways:
* *Versioned migrations* are the classical style, where the migration contents are written by the developer
using the database language.
* *Declarative migrations* are more similar to Infrastructure-as-Code,
where the schema is defined in a Terraform-like language and the migrations commands are
calculated based on the current and desired state of the database. It's possible to use Atlas in
a hybrid way as well, combining both styles, called [Versioned Migration Authoring](https://atlasgo.io/versioned/diff)
where the schema is defined in the Atlas language, but the Atlas engine is used to generate versioned migrations.

> On top of Atlas's ability to load the migration directory as a graph of database entities

Reviewer: ```suggestion
On top of Atlas's ability to load the migration directory as a graph of database entities,
```

> an entire infrastructure of _static code analysis_ was built that is able to provide warnings

Reviewer: ```suggestion
an entire infrastructure of _static code analysis_ was built to provide warnings
```

> about dangerous or inefficient operations, this technique is called **migration linting** and can be

Reviewer: ```suggestion
about dangerous or inefficient operations. This technique is called **migration linting** and can be
```

> integrated with [Atlas Github action](https://atlasgo.io/integrations/github-actions) during CI.

Reviewer: ```suggestion
integrated with the [Atlas Github action](https://atlasgo.io/integrations/github-actions) during CI.
```

In addition, if you would like to run your migrations using Terraform, Atlas has a [Terraform provider](https://atlasgo.io/integrations/terraform-provider)
as well.

> Another key point that Atlas solves is handling migrations integrity, this is a huge problem when working with

Reviewer: ```suggestion
Another key point that Atlas solves is handling migration integrity, which becomes a huge problem when working with
```

> multiple branches that make schema changes, Atlas solves this problem by  using an [Integrity

Reviewer: ```suggestion
multiple branches that all make schema changes. Atlas solves this problem by using an [Integrity
```

file](https://atlasgo.io/concepts/migration-directory-integrity). While we are on the topic of integrity,
one key feature of Atlas is the support for running the migrations inside a transaction, unlike [Goose](https://github.com/pressly/goose/issues/335)

> during the process of migration, Atlas acquires a lock ensuring that only one migration happens at a time

Reviewer: ```suggestion
during the process of migration. Atlas acquires a lock ensuring that only one migration happens at a time
```

> and the migration order/integrity is respected, for cases where problems are found, Atlas makes

Reviewer: ```suggestion
and the migration order/integrity is respected. For cases where problems are found, Atlas makes
```

| Integrity checks | No | No | [Yes](https://atlasgo.io/concepts/migration-directory-integrity) |
| Migration out of order | No | [Possible with hybrid versioning](https://github.com/pressly/goose#hybrid-versioning) | [Possible calculating the directory hash](https://atlasgo.io/versioned/new#recalculating-the-directory-hash) |
| Lock support | Yes | [No](https://github.com/pressly/goose/issues/335) | Yes |
| Use as CLI | Yes | Yes | Yes |
| Use as package | Yes | Yes | Partial support ¹ |
| Versioned Migration Authoring | No | No | [Yes](https://atlasgo.io/versioned/diff) |
| Migration linting | No | No | [Yes](https://atlasgo.io/versioned/lint) |
| Github action | No | No | [Yes](https://atlasgo.io/integrations/github-actions) |
| Terraform provider | No | No | [Yes](https://atlasgo.io/integrations/terraform-provider) |

- 1: Atlas provides [a few packages](https://pkg.go.dev/ariga.io/atlas#section-directories) related to database operations, but the use is limited to complex cases
and there is no package that provides migration usage out of the box.

## Wrapping up

> In this post we saw different strengths of each migration tool, we saw how golang

Reviewer: ```suggestion
In this post we saw different strengths of each migration tool. We saw how Golang
```
